models/modules/unet_generator_attn/unet_generator_attn.py

```diff
@@ -1012,17 +1012,539 @@ def extract(self, a, t, x_shape=(1, 1, 1, 1)):
+        )
+
+        ds = 1
+        for level, mult in enumerate(channel_mults):
+            for _ in range(res_blocks[level]):
+                layers = [
+                    ResBlock(
+                        ch,
+                        self.cond_embed_dim,
+                        dropout,
+                        out_channel=int(mult * inner_channel),
+                        use_checkpoint=use_checkpoint,
+                        use_scale_shift_norm=use_scale_shift_norm,
+                        norm=norm,
+                        efficient=efficient,
+                    )
+                ]
+                ch = int(mult * inner_channel)
+                if ds in attn_res:
+                    layers.append(
+                        AttentionBlockRef(
+                            ch,
+                            use_checkpoint=use_checkpoint,
+                            num_heads=num_heads,
+                            num_head_channels=num_head_channels,
+                            use_new_attention_order=use_new_attention_order,
+                        )
+                    )
+                self.input_blocks_ref.append(EmbedSequentialRef(*layers))
+
+            if level != len(channel_mults) - 1:
+                out_ch = ch
+                self.input_blocks_ref.append(
+                    EmbedSequentialRef(
+                        ResBlock(
+                            ch,
+                            self.cond_embed_dim,
+                            dropout,
+                            out_channel=out_ch,
+                            use_checkpoint=use_checkpoint,
+                            use_scale_shift_norm=use_scale_shift_norm,
+                            down=True,
+                            norm=norm,
+                            efficient=efficient,
+                        )
+                        if resblock_updown
+                        else Downsample(ch, conv_resample, out_channel=out_ch)
+                    )
+                )
+
+        self.middle_block_ref = EmbedSequentialRef(
+            ResBlock(
+                ch,
+                self.cond_embed_dim,
+                dropout,
+                use_checkpoint=use_checkpoint,
+                use_scale_shift_norm=use_scale_shift_norm,
+                norm=norm,
+                efficient=efficient,
+            ),
+            AttentionBlockRef(
+                ch,
+                use_checkpoint=use_checkpoint,
+                num_heads=num_heads,
+                num_head_channels=num_head_channels,
+                use_new_attention_order=use_new_attention_order,
+            ),
+            ResBlock(
+                ch,
+                self.cond_embed_dim,
+                dropout,
+                use_checkpoint=use_checkpoint,
+                use_scale_shift_norm=use_scale_shift_norm,
+                norm=norm,
+                efficient=efficient,
+            ),
+        )
+
```

Comment:
From the paper, it seems there's one additional block of cross-ref attention on the upsample path of the U-Net.
![image](https://github.com/jolibrain/joliGEN/assets/3530657/45d68af8-dd7e-45b9-b327-8ce6781f2cab)